### ISSUE TITLE
Upgrade to Tensorflow2 keras or Pytorch

### ISSUE THREAD

**Issue by smguo** [mehta-lab/microDL#141]
Keras is now part of Tensorflow2's API. Switching to tf.keras will make it easier to maintain the code base and add new models.

**Comment by mattersoflight** [comment]
@smguo , @jennyfolkesson  As we have discussed in meetings, systematic cost-benefit analysis of adopting TF2 or Pytorch is timely. 

We have two strategic goals for the new framework: 1) enable easy training of virtual staining models and 2) enable construction of analysis pipelines that link data conversion, physics-based reconstruction, virtual staining, segmentation, and representation learning. 

The cost we should consider is amount of code that needs to be re-written manually through the end of 2022. The optimal time to start the conversion is sometime in May 2022. Before that, it makes sense to address other issues (update the tests in main branch, #121, #120, #140), and document on this issue exactly which modules need to be converted and how much effort it can take. 

For reconstruction code, we use numpy/cupy, but increasingly using pytorch's auto-diff tools for optimization. The top choices for segmentation models are [AICS segmentation pipeline](https://github.com/AllenCell/aics-ml-segmentation) and [CellPose](https://github.com/MouseLand/cellpose). [Dynamorph](https://github.com/mehta-lab/dynamorph/) and our other representation learning projects also use pytorch. In addition, pytorch marketplace has tools for [continual training](https://github.com/ContinualAI/avalanche) and [model interpretation](https://captum.ai/). It is clear that our long range strategic choice is pytorch. 

However, the most time-efficient approach may be to first migrate to tf.keras 2.0 and then to pytorch, given similarity in their programming models. To estimate the cost accurately and to make plans for how to update the code base, I suggest curating a table of modules that need to be converted, and estimate if they can be [auto-converted](https://www.tensorflow.org/guide/migrate) or need manual refactor when moving to tf.keras or pytorch.

| module |  migration to tf.keras | migration to pytorch | notes |
|--------|----------------------|----------------------|-------|
| micro_dl.networks.UNetStackTo2D| auto refactor | manual refactor | tests will need to be updated/stay the same |

@smguo and @jennyfolkesson if you agree with this approach, please start a google sheet to document the above and any other relevant info.

**Comment by jennyfolkesson** [comment]
I have added modules that need modification in this sheet: https://docs.google.com/spreadsheets/d/1t_sj2fdv5guceN7NqmUHS4Ci266QFB7uxqtk6b0wCVw/edit?usp=sharing

Even if the refactor is "auto" there's usually a lot of manual changes you have to make in my experience. I have no experience with Pytorch so I can't say how much work it would be to manually convert. My assumption is a lot but correct me if I'm wrong.

Regarding main branch tests, I've added a branch called master_tests where I'm updating tests and codebase to work for different use cases 2/2.5/3D, segmentation/regression etc. Some test updates are trivial but a lot are not so I don't have a time estimate for how long it will take yet. Thanks!

**Comment by mattersoflight** [comment]
@Christianfoley and I thought about TF1.x -> TF2.x conversion, and figured that conversion will likely be time consuming and the backward compatibility may be dropped soon. To pilot pytorch, Christian plans to implement a 2.5D UNet in pytorch, train the model with the data we used at the DL course, and compare the metrics of accuracy on test set using both pytorch implementation and Keras 2.1 implementation of 2.5D UNet.  After this test, we can figure the path forward (e.g., how to integrate the pytorch models into microDL pipeline, or build a new pipeline around the pytorch models).
The metrics are described in the "Model Evaluation" section of the paper (https://elifesciences.org/articles/55502). If metrics don't depend on Keras, we should repurpose the implementation from microDL with support from @jennyfolkesson.

**Comment by mattersoflight** [comment]
Closing, due to move to pytorch-lightning.
